FlowCrypt/Controllers/MessageList Extension/MsgListViewController.swift

```diff
@@ -8,6 +8,7 @@
 
 import UIKit
 
+@MainActor
```

Comment:
On a controller, wouldn't this cause some methods that make api calls to run on main? Or as long as we `await` these other methods/services, then the API call would not be on main?

Reply:
It is possible but we should move such methods to services. I mean that view controller is something that interacts with view so we want to be sure that this happens on UI thread. That is why I want to have `@MainActor` on each controller.

Reply:
This is the danger I described earlier - if we just set it on the whole class, things may suddenly run on main that we didn't intend to. I'm not sure it's the best idea without properly studying how @ main actor works, ensuring that api requests are not on main in this situation:

```
@mainactor
class SomeViewController {

  fun handleTap() {
    Task {
       do {
          try await apiClient.someApiCall() // <-- does this run on main or not?
          handleGood()
        } catch {
          handleErr()
        }
    }
  }

  fun handleGood() {
  }

  fun handleErr() {
  }

}
```

Can you please study the docs and also step through this with debugger to make sure?

Reply:
If we run task like this:
```
@MainActor
class SomeViewController {

	func handleTap() {
		print("\(Thread.current): handleTap()")
		Task {
			do {
				print("\(Thread.current): inside Task point 1")
				try await apiClient.someApiCall()
				print("\(Thread.current): inside Task point 2")
				handleGood()
			} catch {
				handleErr()
			}
		}
	}

	func handleGood() {
		print("\(Thread.current): handleGood()")
	}

	func handleErr() {
		print("\(Thread.current): handleErr()")
	}
}
```

Then according to documentation:
> Runs the given nonthrowing operation asynchronously as part of a new top-level task on behalf of the current actor.

So we run task on main thread because we are inside `@MainActor`, but async function may run internally on any other thread. So at "point 1" we are on main thread, but on "point 2" we could be on any other thread, it depends on `apiClient.someApiCall()`. This is the reason why we use `@MainActor`, to be sure that `handleGood()` will be executed on main thread.

On the other hand if we start task like this:
```
@MainActor
class SomeViewController {

	func handleTap() {
		print("\(Thread.current): handleTap()")
		Task.detached {
			do {
				print("\(Thread.current): inside Task point 1")
				try await self.apiClient.someApiCall()
				print("\(Thread.current): inside Task point 2")
				await self.handleGood()
			} catch {
				await self.handleErr()
			}
		}
	}

	func handleGood() {
		print("\(Thread.current): handleGood()")
	}

	func handleErr() {
		print("\(Thread.current): handleErr()")
	}
}
``` 

We have new top level task:
> Runs the given nonthrowing operation asynchronously as part of a new top-level task.

So we run task on another thread (no main), and in "point 1" we are not on main thread. But again, on "point 2" it is possible to be on new thread.

So it depends on how internally works async function.

Reply:
Maybe we need to find out what we want to run on separate thread, like network requests (already implemented in `URLSession` by Apple), our JavaScript functions (Core is actor already), etc.

Reply:
Can we mark these api classes etc with something like BackgroundActor ? :-)

Reply:
I need to check if it is possible to create our own annotations like this.

Reply:
Looks like there is no way to define your own annotations in Swift, so there is no way to have `@BackgroundActor`. On the other hand any user actor by default is "background".

Reply:
got it

Reply:
> Maybe we need to find out what we want to run on separate thread, like network requests (already implemented in URLSession by Apple), our JavaScript functions (Core is actor already), etc.

All API client classes, including Gmail API (which may not be using URLSession - not sure), and Core as you say. IMAP/SMTP client as well. So here the main thing to look into is Gmail and IMAP/SMTP.

Reply:
It's possible to create our own custom actors with `@globalActor` declaration, like
```swift
@globalActor
struct BackgroundActor {
...
}
```

But I think we don't need such `@BackgroundActor` declaration as Swift automatically runs `async/await` operations not on the main thread, so we just need to re-check our API calls to use `async`